File: and-bible/app/src/main/java/net/bible/android/view/activity/page/BibleInfiniteScrollPopulator.java
package net.bible.android.view.activity.page;

import android.util.Log;

import net.bible.android.control.page.CurrentPageManager;

import org.apache.commons.lang3.StringEscapeUtils;

/**
 * Get next or previous page for insertion at the top or bottom of the current webview.
 *
 * @author Martin Denham [mjdenham at gmail dot com]
 * @see gnu.lgpl.License for license details.<br>
 *      The copyright to this program is held by it's author.
 */
public class BibleInfiniteScrollPopulator {

	private final BibleViewTextInserter bibleViewtextInserter;
	private final CurrentPageManager currentPageManager;

	private static final String TAG = "BibleInfiniteScrollPop";

	public BibleInfiniteScrollPopulator(BibleViewTextInserter bibleViewtextInserter, CurrentPageManager pageManager) {
		this.bibleViewtextInserter = bibleViewtextInserter;
		currentPageManager = pageManager;
	}

	public void requestMoreTextAtTop(String textId) {
		Log.d(TAG, "requestMoreTextAtTop");
		//TODO do in background thread
		// get page fragment for previous chapter
		String fragment = currentPageManager.getCurrentPage().getPreviousPageFragment();
		fragment = StringEscapeUtils.escapeEcmaScript(fragment);
		bibleViewtextInserter.insertTextAtTop(textId, fragment);
	}

	public void requestMoreTextAtEnd(String textId) {
		Log.d(TAG, "requestMoreTextAtEnd");
		//TODO do in background thread
		// get page fragment for previous chapter
		String fragment = currentPageManager.getCurrentPage().getNextPageFragment();
		fragment = StringEscapeUtils.escapeEcmaScript(fragment);
		bibleViewtextInserter.insertTextAtEnd(textId, fragment);
	}
}

File: and-bible/app/src/main/java/net/bible/android/view/activity/page/BibleInfiniteScrollPopulator.kt
package net.bible.android.view.activity.page

import net.bible.android.control.page.CurrentBiblePage
import net.bible.android.control.page.CurrentPageManager
import net.bible.android.view.activity.base.Callback

import org.apache.commons.lang3.StringEscapeUtils
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.debug
import org.jetbrains.anko.doAsync

/**
 * Get next or previous page for insertion at the top or bottom of the current webview.
 *
 * @author Martin Denham [mjdenham at gmail dot com]
 * @see gnu.lgpl.License for license details.<br></br>
 * The copyright to this program is held by it's author.
 */
class BibleInfiniteScrollPopulator(private val bibleViewtextInserter: BibleViewTextInserter, private val currentPageManager: CurrentPageManager) : AnkoLogger {

    fun requestMoreTextAtTop(chapter: Int, textId: String, callback: Callback) {
        debug("requestMoreTextAtTop")
        // do in background thread
        doAsync {
            // get page fragment for previous chapter
            val currentPage = currentPageManager.currentPage
            if (currentPage is CurrentBiblePage) {
                var fragment = currentPage.getFragmentForChapter(chapter)
                fragment = StringEscapeUtils.escapeEcmaScript(fragment)
                bibleViewtextInserter.insertTextAtTop(textId, fragment)
            }
            // tell js interface that insert is complete
            callback.okay()
        }
    }

    fun requestMoreTextAtEnd(chapter: Int, textId: String) {
        debug("requestMoreTextAtEnd")
        // do in background thread
        doAsync {
            // get page fragment for previous chapter
            val currentPage = currentPageManager.currentPage
            if (currentPage is CurrentBiblePage) {
                var fragment = currentPage.getFragmentForChapter(chapter)
                fragment = StringEscapeUtils.escapeEcmaScript(fragment)
                bibleViewtextInserter.insertTextAtEnd(textId, fragment)
            }
        }
    }
}
